cranelift-codegen/src/isa/x86/fde.rs

```diff
@@ -0,0 +1,354 @@
+//! Support for FDE data generation.
+
+use crate::binemit::{FrameUnwindOffset, FrameUnwindSink, Reloc};
+use crate::ir::{FrameLayoutChange, Function};
+use crate::isa::{CallConv, RegUnit, TargetIsa};
+use alloc::vec::Vec;
+use core::convert::TryInto;
+use gimli::write::{
+    Address, CallFrameInstruction, CommonInformationEntry, EhFrame, EndianVec,
+    FrameDescriptionEntry, FrameTable, Result, Writer,
+};
+use gimli::{Encoding, Format, LittleEndian, Register, X86_64};
+use std::ptr;
+
+pub type FDERelocEntry = (FrameUnwindOffset, Reloc);
+
+const FUNCTION_ENTRY_ADDRESS: Address = Address::Symbol {
+    symbol: 0,
+    addend: 0,
+};
+
+#[derive(Clone)]
+struct FDEWriter {
+    vec: EndianVec<LittleEndian>,
+    relocs: Vec<FDERelocEntry>,
+}
+
+impl FDEWriter {
+    fn new() -> Self {
+        Self {
+            vec: EndianVec::new(LittleEndian),
+            relocs: Vec::new(),
+        }
+    }
+    fn into_vec_and_relocs(self) -> (Vec<u8>, Vec<FDERelocEntry>) {
+        (self.vec.into_vec(), self.relocs)
+    }
+}
+
+impl Writer for FDEWriter {
+    type Endian = LittleEndian;
+    fn endian(&self) -> Self::Endian {
+        LittleEndian
+    }
+    fn len(&self) -> usize {
+        self.vec.len()
+    }
+    fn write(&mut self, bytes: &[u8]) -> Result<()> {
+        self.vec.write(bytes)
+    }
+    fn write_at(&mut self, offset: usize, bytes: &[u8]) -> Result<()> {
+        self.vec.write_at(offset, bytes)
+    }
+    fn write_address(&mut self, address: Address, size: u8) -> Result<()> {
+        match address {
+            Address::Constant(_) => self.vec.write_address(address, size),
+            Address::Symbol { .. } => {
+                assert_eq!(address, FUNCTION_ENTRY_ADDRESS);
+                let rt = match size {
+                    4 => Reloc::Abs4,
+                    8 => Reloc::Abs8,
+                    _ => {
+                        panic!("Unexpected address size at FDEWriter::write_address");
+                    }
+                };
+                self.relocs.push((self.vec.len().try_into().unwrap(), rt));
+                self.vec.write_udata(0, size)
+            }
+        }
+    }
+}
+
+fn return_address_reg(isa: &dyn TargetIsa) -> Register {
```

Comment:
Also on the point of being architecture agnostic, I'd specifically wanted `DwarfRegMapper` to clearly indicate how to add support for non-x86_64 architectures. You mentioned wanting `cranelift-module` changed for other reasons, is there something unsuitable about `struct DwarfRegMapper` in #902? If it was sufficient, and we're just deferring non-x86 thought for the moment, I'd like to revive `DwarfRegMapper` in a subsequent PR.